superset/viz.py

```diff
@@ -63,22 +65,42 @@ class BaseViz(object):
     cache_type = 'df'
     enforce_numerical_metrics = True
 
-    def __init__(self, datasource, form_data, force=False):
-        if not datasource:
-            raise Exception(_('Viz is missing a datasource'))
+    def __init__(self, datasource, form_data, query_obj={}, force=False):
+        if datasource:
+            self.datasource = datasource
+        else:
+            datasource_dict = query_obj.get('datasource', {})
+            ds = ConnectorRegistry.get_datasource(
+                datasource_dict.get('type'), datasource_dict.get('id'), db.session)
+            if not ds:
+                raise Exception(_('Viz is missing a datasource'))
+            else:
```

Comment:
NIT: no need for an else when the if clause exits

Reply:
👍